File: Python3/1 - 200/116 Populating Next Right Pointers in Each Node.py
# 56ms, 95.53%; 14.3MB, 100.00%
"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""
class Solution:
    def connect(self, root: 'Node') -> 'Node':
        if not root: return None
        st = [root]
        while True:
            if st[0].left is None: break
            _new = []
            while st:
                node = st.pop(0)
                _new.extend([node.left, node.right])
            for i in range(1, len(_new)):
                _new[i - 1].next = _new[i]
            st = _new
        return root

File: 1 - 200/116 Populating Next Right Pointers in Each Node.cc
// 20ms, 93.65%; 19.5MB, 100.00%
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};
*/
class Solution
{
public:
    Node *connect(Node *root)
    {
        if (!root) return nullptr;
        queue<Node *> q;
        q.push(root);
        while (true) {
            if (!q.front()->left) break;
            for (int i = 0, size = q.size(); i < size; ++i) {
                if (i) q.back()->next = q.front()->left;
                q.push(q.front()->left);
                q.back()->next = q.front()->right;
                q.push(q.front()->right);
                q.pop();
            }
        }
        return root;
    }
};
